botocore/utils.py

```diff
@@ -260,70 +265,108 @@ def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
             proxies=get_environ_proxies(self._url),
         )
 
-    def _get_request(self, url, timeout, num_attempts=1):
+    def retrieve_iam_role_credentials(self):
+        try:
+            role_name = self._get_iam_role()
+            credentials = self._get_credentials(role_name)
```

Comment:
I think the retry should trigger a restart of the *entire* process, including looking up the role name.  Right now we assume that if we get the role_name, we only need to retry if we aren't able to get credentials.  This won't handle the case where we get a bad role name and then try to retrieve credentials (which will never work).

Reply:
I was thinking about that as well, but I decided not to do that in order to keep parity with the existing behavior where we only do retries on a per IMDS request basis. I do not think we want to make the general assumption to refetch the role name on retryable credential grabs. For example when you either swap a role or disassociate a role, it is possible to get a 404 response back for a credential request. For the case where the role was completely disassociated, it is not appropriate to retry the entire process as there are no more credentials available. For role swaps, it would be appropriate to retry the entire process. The big problem is that it is impossible to disambiguate the 404's between both scenarios.

Reply:
Following up from offline discussion:  we should keep this logic as you have it.  It keeps the logic closer to the previous behavior.  While it's possible that we miss a case where we have a role name that is completely invalid, this realistically only affects the initial request to fetch credentials.  Once botocore has credentials, any refresh attempts during the advisory refresh is not a hard failure.  This means in order for this to be a hard failure we'd need to repeatedly see this failure which would require the role's repeatedly being switched while botocore is fetching credentials.